src/github.com/matrix-org/dendrite/roomserver/storage/membership_table.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"database/sql"
+
+	"github.com/matrix-org/dendrite/roomserver/types"
+)
+
+type membershipState int64
+
+const (
+	membershipStateLeaveOrBan membershipState = 0
```

Comment:
I don't like `0` enums because it is the zero-value for ints. It's too easy to make a mistake here, when we have usages like:
```go
return u.membership == membershipStateLeaveOrBan
```
A zero-value `membershipUpdater` would return `true` here, whereas really it should be failing (fast preferably).